### ISSUE TITLE
Change keyserver to hkps://hkps.pool.sks-keyservers.net

### ISSUE THREAD

**Issue by ypid** [debops/ansible-core#21]
Prerequisites:

- [ ] Configure Root CA for apt-key

Refs:

* https://gpgtools.tenderapp.com/discussions/problems/28674-hkps
* https://www.privacy-handbuch.de/handbuch_32s.htm (German)

**Comment by drybjed** [comment]
I'm not entirely sure that going through the pool over HKPS directly is a good idea. If you look at the [SKS pool status page](https://sks-keyservers.net/status/), you can see that many pool mirrors don't support encrypted connections at all. Furthermore, even if you connect over HTTPS/HKPS initially, the pool redirector can just redirect you to a HTTP/HPKP connection, in which case encryption is pointless anyway.

When you read a [related thread on Security StackExchange](https://security.stackexchange.com/questions/4161/shouldnt-gpg-key-fetching-use-a-secure-connection), OpenPGP security doesn't rely on secure connections to the keyservers at all - the keys themselves are the source of trust. In which case wether you connect to the keyservers over HTTP or HTTPS doesn't really matter all that much. If you are concerned about somebody finding out what keys you are getting, I have two solutions for you, which are already implemented in DebOps.

First one, is to select a specific keyserver mirror from the above list of SKS keyservers, which supports encrypted connections, and configure it in the `debops.core` role as the default keyserver. For example:
```
core__keyserver: 'hkps://keyserver.example.com'
```
This way, all roles that know about the `debops.core` and use its facts, will be automatically configured to use that specific server. If you make sure that it uses a trusted certificate, for example signed by Let's Encrypt, you don't need to do anything else.

Another option would be to set up your own local keyserver, for example with the [debops.sks](https://github.com/debops/ansible-sks/) role. Then, you can pull the whole keyring from a nearest keyserver (I think ~5GB, not sure at the moment) and update it periodically. After that, nobody will know which keys you wanted, since you grabbed all of them anyway. And you can setup your own internal certificates on that keyserver, so that all of your hosts trust it implicitly, still no issues with the random untrusted certificates.

I suppose that it is more work than messing around with `apt-key` and relying on some bugs to make it work, but at least for me it would be more trustworthy than adding some random Root Certificate Authority on other people's hosts by default. Remember how DebOps is used.

**Comment by ypid** [comment]
> Furthermore, even if you connect over HTTPS/HKPS initially, the pool redirector can just redirect you to a HTTP/HPKP connection, in which case encryption is pointless anyway.

I don’t think that is how hkps://hkps.pool.sks-keyservers.net works.

> I suppose that it is more work than messing around with apt-key and relying on some bugs to make it work, but at least for me it would be more trustworthy than adding some random Root Certificate Authority on other people's hosts by default. Remember how DebOps is used.

Fully agreed. A proper solution would be needed obviously. Also, the Root CA MUST not go into the system ca store as it is not needed. It could be configured in a gpg.conf and this config could be specified for apt_key if support would be added.

**Comment by ypid** [comment]
To add a bit more why I think this would be a good idea. [This comment](https://github.com/QubesOS/qubes-issues/issues/2240#issuecomment-239650738) got me scared :wink:

**Comment by ypid** [comment]
To elaborate a bit more on this, the concept is called "defense in depth". I understand that the key fingerprint should be enough but that relies on a few assumptions which might not always be resilient. Which also applies in this case (ref: [apt_key module does not verify key fingerprints](https://github.com/ansible/ansible-modules-core/issues/5237)).

Want to rethink your opinion?

**Comment by drybjed** [comment]
Agreed, this is a pretty serious issue, especially that short OpenPGP key ids are easily spoofable these days. I hope that the Ansible team will take the above issue seriously.

OK then, what can we do on our end? Since putting a random Root Certificate Authority in the system CA store is out of the question, we could download the RootCA certificate on each host over HTTPS and  store it in a known location. But:

- `apt-key` does not support using specific Root CA, at least not without using a buggy code
- Ansible `apt_key` module does not support using a specific Root CA either, and this is the mechanism DebOps uses to manage the APT GPG keys.

Even if you add support for specific Root CA in the `apt-key` upstream, it won't get into Debian Jessie and will be available most likely in Debian Buster, because it's too late for this improvement to get into Debian Stretch before the freeze on 5th of November (remember, remember).

It is definitely a good idea to improve `apt-key` to support this scenario - I'm not against that. I currently don't see a way to do what you are proposing in a sane way in the environment that DebOps exists in. Therefore I would still suggest to use the other methods I explained above - either specify a keyserver via the `debops.core` variable, and you can choose whichever HTTPS keyserver you want, or setup and use your own keyserver using the `debops.sks` role.
